pennylane_lightning/lightning_qubit.py

```diff
@@ -441,17 +353,97 @@ def compute_vjp(self, dy, jac, num=None):
         )
         return vjp_tensor
 
+    def vjp(self, tape, dy, starting_state=None, use_device_state=False):
+        """Generate the processing function required to compute the vector-Jacobian products of a tape.
+        Args:
+            tape (.QuantumTape): quantum tape to differentiate
+            dy (tensor_like): Gradient-output vector. Must have shape
+                matching the output shape of the corresponding tape.
+        Keyword Args:
+            starting_state (tensor_like): post-forward pass state to start execution with. It should be
+                complex-valued. Takes precedence over ``use_device_state``.
+            use_device_state (bool): use current device state to initialize. A forward pass of the same
+                circuit should be the last thing the device has executed. If a ``starting_state`` is
+                provided, that takes precedence.
+        Returns:
+            The processing function required to compute the vector-Jacobian
+            products of a tape.
+        """
+        if self.shots is not None:
+            warn(
+                "Requested adjoint differentiation to be computed with finite shots."
+                " The derivative is always exact when using the adjoint differentiation method.",
+                UserWarning,
+            )
+
+        num_params = len(tape.trainable_params)
+
+        if num_params == 0:
+            return lambda _: None
+
+        if math.allclose(dy, 0):
+            return lambda _: math.convert_like(np.zeros([num_params]), dy)
+
+        # To support np.complex64 based on the type of self._state
+        dtype = self._state.dtype
+        if dtype == np.complex64:
+            use_csingle = True
+        elif dtype == np.complex128:
+            use_csingle = False
+        else:
+            raise TypeError(f"Unsupported complex Type: {dtype}")
+
+        if use_csingle:
+            V = VectorJacobianProductC64()
+        else:
+            V = VectorJacobianProductC128()
+
```

Comment:
Should we combine these two `if` statements, since `dtype` isn't used outside of this section?

Reply:
:100: Fixed it!